File: Lazily_Propogated/CPP/Max_Subarray_Sum_Query_only.cpp

class SegmentTree{
    vector<tuple<int,int,int>> tree;
    int n;
    int n1;
    public:

        tuple<int,int,int> func(tuple<int,int,int> &a,tuple<int,int,int> &b){
            tuple<int,int,int> tpnew;
            tpnew = {min(get<0>(a),get<0>(b)),max(get<1>(a),get<1>(b)),max(max(get<2>(a),get<2>(b)),get<1>(b)-get<0>(a))};
            return tpnew;
        }

        SegmentTree(vi L){
            n = len(L);
            n1 = 1;
            while (n1 < n) n1 <<= 1;
            tree.assign(2*n1, {INF, -INF, -INF});
            
            for(int i=0;i<len(L);i++){
                tree[i+n1] = {L[i],L[i],0};
            }

            for(int i=n1-1;i>0;i--){
                tree[i] = func(tree[2*i],tree[2*i+1]);
            }
        }

        tuple<int,int,int> query(int l,int r){
            l += n1;
            r += n1;
            tuple<int,int,int> left = {INF,-INF,-INF};
            tuple<int,int,int> right = {INF,-INF,-INF};
            while(l<=r){
                if(l&1){
                    left = func(left,tree[l]);
                    l++;
                }
                if(!(r&1)){
                    right = func(tree[r],right);
                    r--;
                }
                l >>= 1;
                r >>= 1;
            }
            return func(left,right);
        }
};

File: Lazily_Propogated/Max_Subarray_Sum_Query_only.py
class SegmentTree:
    @staticmethod
    def func(left, right):
        return (min(left[0], right[0]),max(left[1], right[1]),max(left[2], right[2], right[1] - left[0]))
    def __init__(self, arr):
        arr = [0] + arr
        for i in range(1, len(arr)):
            arr[i] += arr[i - 1]
        self.n1 = len(arr) - 1
        self.n = (1 << ((len(arr) - 1).bit_length()))
        self.tree = [(float('inf'), -float('inf'), -float('inf'))] * (2 * self.n)
        for i in range(len(arr)):
            self.tree[self.n + i] = (arr[i], arr[i], 0)
        for i in range(self.n - 1, 0, -1):
            self.tree[i] = self.func(self.tree[2 * i], self.tree[2 * i + 1])
    def query(self, l, r):
        """Query range [l, r] inclusive."""
        l += self.n
        r += self.n
        res_left = (float('inf'), -float('inf'), -float('inf'))
        res_right = (float('inf'), -float('inf'), -float('inf'))
        while l <= r:
            if l % 2 == 1:
                res_left = self.func(res_left, self.tree[l])
                l += 1
            if r % 2 == 0:
                res_right = self.func(self.tree[r], res_right)
                r -= 1
            l //= 2
            r //= 2
        return self.func(res_left, res_right)